### ISSUE TITLE
Docker: Publish hostname of the service on the local network

### ISSUE THREAD

**Issue by nadigo** [photoprism/photoprism#2470]
As a user who installed photoprism on docker with macvaln network, I would like to use photoprism.local:port to access my site.

* missing MDNS / Avahi support on docker Image

**Comment by lastzero** [comment]
Is this a common best practice? Please add more details and examples, ideally based on our feature request template which we provide for this.

**Comment by nadigo** [comment]
yes, to access .local domains, the host has to publish its hostname over the local network using mDNS / Bonjour, the most common package for Debian is avahi-daemon. ([Avahi]https://en.wikipedia.org/wiki/Avahi_(software)) is a free zero-configuration implementation of the mDNS protocol.

Running avahi on a none privileged container needs one configuration since avahi require dbus (but can run without it).
You need to disable dbus "enable-dbus=no" on the avahi-daemon config file "/etc/avahi/avahi-daemon.conf".

see [here](https://gnanesh.me/avahi-docker-non-root.html)  

```
# Dockerfile

RUN apt-get update && apt-get install -y --no-install-recommends avahi-daemon 
RUN printf "[server]\nenable-dbus=no\n" >> /etc/avahi/avahi-daemon.conf \
&& service avahi-daemon restart 
```

If you direct me to where in the photoprism git you install packages, I can have a look and use your way to add packages and add avahi-daemon similarly, then submit a pull request. 

On [my version](https://hub.docker.com/layers/237821395/nadigo/photoprism/latest/images/sha256-292ffdf07fa35e22c60566f88390f1ca3b936b62cd378055da33867b4c4b89d7?context=repo) of the photoprism container  I added manually the avahi-daemon packages similar to the above and it works fine. 

Let me know.

**Comment by lastzero** [comment]
Could we implement this directly in the app or is it required to use this daemon? Maybe there is a Go library we could use instead. A bit of research on alternatives would be amazing.

**Comment by nadigo** [comment]
There is also a go package I was looking at.
I’ll have another look and post it here.

**Comment by nadigo** [comment]
I’ll have a closer look any post it here.

**Comment by nadigo** [comment]
[this](https://pkg.go.dev/github.com/hashicorp/mdns#section-documentation) looks like small and simple one
you need only the server part

note that the service string has to include `_http._tcp` and the actual `$IP` and `$PORT` from the os / config.

`service, _ := mdns.NewMDNSService(host, "_http._tcp", "", "", $PORT, $IP, info)`

**Comment by lastzero** [comment]
Could this have security or other implications, especially when users are running the same image on public cloud servers or have multiple instances running on the same server? If so, what solution(s) do you propose?

**Comment by gyto6** [comment]
Computers hostname are not accessible from WAN or IPV6 higher-sub-tunnel, it's only accessible within the same network neighborhood, explaining the following .local.

Furthermore, in public cloud instance, each server (physical or virtual) is compartmentalized in its own network.
I don't see any security breach or troubleshooting source as hostnames are not unique and only available in a LAN.

It's different from a Domain Name set on a registrar which is linked to an IP from WAN and can precisely identify a specific target on the net.

**Comment by nadigo** [comment]
Agree. since photoprism is designed for local / house networks it should be
broadcasting its service and hostname via mDNS like most appliances do.

**Comment by lastzero** [comment]
How does the Docker container know which IP to publish? Containers usually only have an internal IP unless a bridged network has been configured, which is not the default. If you can only use it locally anyway, there's probably not much to publish since the hostname will always be localhost?

**Comment by nadigo** [comment]
In my mind the typical architacture is photoprism hosted on a local NAS
(natively or most likely via docker) and an end-user accessing it via a
browser from a different computer over the lan.
If that's the use case you had in mind, the default docker network should
be 'bridge'.

**Comment by lastzero** [comment]
Wouldn't a bridge network require DHCP and its own IP? I'm not sure we should expect that. And I don't expect many users to use it that way, especially with Synology and similar consumer NAS devices. Of course, I could be wrong...

**Comment by nadigo** [comment]
Why use phtooprism **unless** you are remote from your image repository ....
(We can take it for granted that DHCP will assign IP over the lan)

**Comment by lastzero** [comment]
People often just use it on their local computer or access their instance over the internet. In both cases a local domain doesn't help much. Likely, there is also no easy way to get a signed TLS certificate for a local domain. It's important to understand the exact use case. So from what I understand it mostly makes sense to publish a name if the IP is private but not in a range that Docker uses internally?

**Comment by nadigo** [comment]
agreed, to publish photoprirm over the internet you would need a lan IP + port fw + dynamic or fix DNS. So bridge networking is a basic config for all use cases. The broadcasting of the local hostname IP over mDNS has no impact on remote access as most routers will not pass the broadcast. I use many home network services, many of them on PI and some on NAS or other home servers **almost all of them use mDNS.**

**Comment by nadigo** [comment]
to sum up, I think bridge networking + mDNS should be the default + adding `hostname: photoprism` to docker and changing the documentation to let users know that  photoprism should be accessed on http://photoprism.local:PORT

**Comment by harshavmb** [comment]
Hi @nadigo,

I think your photoprism host can also publish the hostname right? Is there any specific reason, you want this to be published by the application running on the host? If multiple applications running on the host publish different A-record names with `.local` TLD, wouldn't it be confusing?

**Comment by gyto6** [comment]
> Hi @nadigo,
> 
> I think your photoprism host can also publish the hostname right? Is there any specific reason, you want this to be published by the application running on the host? If multiple applications running on the host publish different A-record names with `.local` TLD, wouldn't it be confusing?

An host should have it's own domain name. It must not be mixed with virtualized guest. 

Most users are not aware about DHCP or fixed IP, and that's where DNS is supposed to help management.
